synapse/appservice/api.py

```diff
@@ -231,13 +231,26 @@ async def push_bulk(
                 json_body=body,
                 args={"access_token": service.hs_token},
             )
+            logger.debug(
+                "push_bulk to %s succeeded! events=%s",
+                uri,
+                [event.get("event_id") for event in events],
+            )
             sent_transactions_counter.labels(service.id).inc()
             sent_events_counter.labels(service.id).inc(len(events))
             return True
         except CodeMessageException as e:
-            logger.warning("push_bulk to %s received %s", uri, e.code)
+            logger.warning(
+                "push_bulk to %s received code=%s msg=%s", uri, e.code, e.msg
+            )
         except Exception as ex:
-            logger.warning("push_bulk to %s threw exception %s", uri, ex)
+            logger.warning(
+                "push_bulk to %s threw exception(%s) %s args=%s",
+                uri,
+                type(ex).__name__,
+                ex,
+                ex.args,
+            )
```

Comment:
Without these extra params, it's pretty much impossible to debug errors like:

```
push_bulk to http://host.docker.internal:9111/transactions/1 threw exception 400 Bad Request

push_bulk to http://localhost:9000/transactions/1 threw exception An error occurred while connecting: 99: Cannot assign requested address.

push_bulk to http://0.0.0.0:9111/transactions/1 threw exception Connection was refused by other side: 111: Connection refused.

push_bulk to http://host.docker.internal:9111/transactions/1 received 504
```

When there is more info available like:
```
push_bulk to http://host.docker.internal:9111/transactions/1 threw PartialDownloadError code=b'400' msg=b'Bad Request' response=b'Client sent an HTTP request to an HTTPS server.\n'

push_bulk to http://localhost:9111/transactions/1 threw exception(ConnectError) An error occurred while connecting: 99: Cannot assign requested address. args=('Cannot assign requested address',)

push_bulk to https://host.docker.internal:9111/transactions/1 threw exception(RequestTransmissionFailed) [<twisted.python.failure.Failure OpenSSL.SSL.Error: [('SSL routines', 'tls_process_server_certificate', 'certificate verify failed')]>]
```

Reply:
An alternate approach here is to do:

```python
logger.warning(
    "push_bulk to %s threw exception %s", 
    uri, 
    ex,
    exc_info=logger.isEnabledFor(logging.DEBUG),
)
```

which should include exception info if the logging level is set to debug. Your method also works

Reply:
Maybe both 👀. The warnings are kinda useless without the little extra info on type of error and arguments. But also useful to have the stack when debugging.

examples with `exc_info=True`
```python
push_bulk to http://host.docker.internal:65302/transactions/1 received code=500 msg=Internal Server Error
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/appservice/api.py", line 229, in push_bulk
    await self.put_json(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/client.py", line 639, in put_json
    raise HttpResponseException(
synapse.api.errors.HttpResponseException: 500: Internal Server Error
```

```python
push_bulk to http://host.docker.internal:49443/transactions/1 threw exception(ConnectionRefusedError) Connection was refused by other side: 111: Connection refused. args=('Connection refused',)
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/synapse/appservice/api.py", line 229, in push_bulk
    await self.put_json(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/client.py", line 630, in put_json
    response = await self.request(
  File "/usr/local/lib/python3.8/site-packages/synapse/http/client.py", line 440, in request
    response = await make_deferred_yieldable(request_deferred)
twisted.internet.error.ConnectionRefusedError: Connection was refused by other side: 111: Connection refused.
```

Reply:
Fair. Ideally converting an exception to a string should give us useful information, but clearly those exceptions (which we don't control) don't follow that rule :/